[Cocoa] Add a heuristic to detect and sample colors in fixed-position elements at the edges of the viewport

whsieh opened the pull request:
#### f5aa57b6b88aa2ffca5cd1457a45c56723031bcd
<pre>
[Cocoa] Add a heuristic to detect and sample colors in fixed-position elements at the edges of the viewport
<a href="https://bugs.webkit.org/show_bug.cgi?id=287690">https://bugs.webkit.org/show_bug.cgi?id=287690</a>
<a href="https://rdar.apple.com/144839983">rdar://144839983</a>

Reviewed by Abrar Rahman Protyasha.

Work towards <a href="https://rdar.apple.com/144839983">rdar://144839983</a>; see below for more details.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::fixedContainerEdges const):

Add a helper method that returns the set of `FixedContainerEdges` (see the new header below), by
first hit-testing around all 4 edges of the layout viewport in search of elements with fixed-
position ancestors in the render tree; for each edge with a fixed ancestor, we then use the new
page color sampling helper below to extract the predominant color.

* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/PageColorSampler.cpp:
(WebCore::PageColorSampler::predominantColor):

Add a helper method that samples the predominant color in the given rect (in absolute coordinates)
by taking a snapshot, building a histogram of color frequencies, and checking whether the number of
similar colors represents a majority of all sampled colors.

* Source/WebCore/page/PageColorSampler.h:
* Source/WebCore/platform/FixedContainerEdges.h: Added.

Add a helper class, `FixedContainerEdges`, that represents two pieces of information about each of
the 4 edges of the layout viewport (top, left, bottom, right):

-   Whether there&apos;s a fixed-position container near that edge.
-   A sampled predominant color, or the invalid color if none is detected (e.g. due to too many
    different colors).

Canonical link: <a href="https://commits.webkit.org/290411@main">https://commits.webkit.org/290411@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39503e57c6c7757e4c38e8e89751474d59911e0d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89909 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44825 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94909 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40682 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91961 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9825 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17719 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69234 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26846 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7520 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81584 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49590 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7246 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35961 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39816 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77589 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37005 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96732 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17096 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12550 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78169 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17352 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77403 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77428 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19127 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21869 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20474 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10285 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17107 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22432 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16848 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20300 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18631 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/PageColorSampler.cpp:
```diff
@@ -274,4 +277,95 @@ std::optional<Color> PageColorSampler::sampleTop(Page& page)
+        if (a.second == b.second)
+            return a.first.luminance() > b.first.luminance(); // Break ties deterministically.
+        return a.second > b.second;
+    });
```
Did you consider `std::stable_sort`? It would not bubble up equivalent frequencies with higher luminance, but it would be deterministic in that it would preserve the relative order of equally frequent colors from before the sort.

aprotyas replied:
but maybe you want to give higher priority to higher luminance

whsieh replied:
Ah, yeah — I think `std::stable_sort` is really what I want (I'm not sure it's really important to break ties by _luminance_ — more that ties are broken in the same way so as to be more robust against predominant color thrashing)

aprotyas commented on Source/WebCore/page/PageColorSampler.cpp:
```diff
@@ -274,4 +277,95 @@ std::optional<Color> PageColorSampler::sampleTop(Page& page)
+    std::optional<Color> mostFrequentColor;
+    unsigned mostFrequentColorCount = 0;
+
+    for (auto& [color, count] : colorsByDescendingFrequency) {
+        if (!mostFrequentColor) {
+            mostFrequentColor = color;
+            mostFrequentColorCount = count;
+            continue;
+        }
+
+        if (!colorsAreSimilar(*mostFrequentColor, color))
+            continue;
+
+        mostFrequentColorCount += count;
+
+        if (mostFrequentColorCount > minimumSampleCountForPredominantColor)
+            return WTFMove(*mostFrequentColor);
+    }
```
Small remark, consider the priority list `{ { A, 5 }, { B, 3 }, { C, 2 }, { D, 1 } }` where B, C, and D are all similar to each other. In that case we would want to return `B`, right? But I think the way the algorithm is written now, we would favor `A` instead.

That's probably fine though

whsieh replied:
Yeah — this is a known issue with this approach. I considered making it more robust by aggregating _all_ similar colors (not just the first one), but doing so would result in either O(n^2) perf (n being sample count), or require building some kind of spatial tree to (essentially) run k-neighbors, which seemed a bit excessive for this corner case.

I'll add a FIXME mentioning this, but keep it as-is for now (since I think this simpler approach will work well in practice).

aprotyas replied:
Makes sense, I doubt it will end up mattering.

aprotyas replied:
k-neighbours...
![image](https://github.com/user-attachments/assets/d8f9b4e6-2738-46af-bdc8-47e74ca92e50)

aprotyas commented on Source/WebCore/page/PageColorSampler.h:
```diff
@@ -35,6 +35,7 @@ class Page;
 class PageColorSampler {
 public:
     static std::optional<Color> sampleTop(Page&);
+    static Color predominantColor(Page&, const LayoutRect&);
```
You probably want to forward decl `LayoutRect`

whsieh replied:
Will fix!

whsieh committed:
[Cocoa] Add a heuristic to detect and sample colors in fixed-position elements at the edges of the viewport
https://bugs.webkit.org/show_bug.cgi?id=287690
rdar://144839983

Reviewed by Abrar Rahman Protyasha.

Work towards rdar://144839983; see below for more details.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::fixedContainerEdges const):

Add a helper method that returns the set of `FixedContainerEdges` (see the new header below), by
first hit-testing around all 4 edges of the layout viewport in search of elements with fixed-
position ancestors in the render tree; for each edge with a fixed ancestor, we then use the new
page color sampling helper below to extract the predominant color.

* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/PageColorSampler.cpp:
(WebCore::PageColorSampler::predominantColor):

Add a helper method that samples the predominant color in the given rect (in absolute coordinates)
by taking a snapshot, building a histogram of color frequencies, and checking whether the number of
similar colors represents a majority of all sampled colors.

* Source/WebCore/page/PageColorSampler.h:
* Source/WebCore/platform/FixedContainerEdges.h: Added.

Add a helper class, `FixedContainerEdges`, that represents two pieces of information about each of
the 4 edges of the layout viewport (top, left, bottom, right):

-   Whether there's a fixed-position container near that edge.
-   A sampled predominant color, or the invalid color if none is detected (e.g. due to too many
    different colors).

Canonical link: https://commits.webkit.org/290411@main
